File: CodeChef/Chef_Builds_Twitter.py
#https://www.codechef.com/problems/HRYQX01?tab=statement

import heapq
from collections import defaultdict

class Twitter:

    def __init__(self):
        self.time = 0
        self.tweets = defaultdict(list)
        self.following = defaultdict(set)
        
    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweets[userId].append((self.time, tweetId))
        self.time += 1
        
    def getNewsFeed(self, userId: int):
        res = []
        heap = []
        
        #push latest tweet
        if self.tweets[userId]:
            idx = len(self.tweets[userId]) - 1
            t, tweetId = self.tweets[userId][idx]
            heapq.heappush(heap, (-t, tweetId, idx, userId))
        
        for f in self.following[userId]:
            if self.tweets[f]:
                idx = len(self.tweets[f]) - 1
                t, tweetId = self.tweets[f][idx]
                heapq.heappush(heap, (-t, tweetId, idx, f))
                
        while heap and len(res) < 10:
            t, tweetId, idx, u = heapq.heappop(heap)
            res.append(tweetId)
            
            if idx > 0:
                next_time, next_tweet = self.tweets[u][idx - 1]
                heapq.heappush(heap, (-next_time, next_tweet, idx - 1, u))
        
        return res
        
    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId != followeeId:
            self.following[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        self.following[followerId].discard(followeeId)
        

File: CodeChef/Chef_Builds_Twitter.cpp
//https://www.codechef.com/problems/HRYQX01?tab=statement

class Twitter {
    private: long long initTime;
    vector < vector < pair < long long,
    int >>> tweets;
    vector < unordered_set < int >> following;

    public: Twitter() {
        // write your code here 
        initTime = 0;
        tweets.resize(501);
        following.resize(501);

    }

    void postTweet(int userId, int tweetId) {
        // write your code here 
        tweets[userId].push_back({
            initTime++,
            tweetId
        });

    }

    vector < int > getNewsFeed(int userId) {
        // write your code here 
        vector < int > res;
        priority_queue < tuple < long long, int, int, int >> pq;

        if (!tweets[userId].empty()) {
            int i = tweets[userId].size() - 1;
            auto & t = tweets[userId][i];
            pq.push({
                t.first,
                t.second,
                i,
                userId
            });
        }

        for (int f: following[userId]) {
            if (!tweets[f].empty()) {
                int i = tweets[f].size() - 1;
                auto & t = tweets[f][i];
                pq.push({
                    t.first,
                    t.second,
                    i,
                    f
                });
            }
        }

        while (res.size() < 10 && pq.empty() == 0) {
            auto[time, tweetId, idx, u] = pq.top();
            pq.pop();

            res.push_back(tweetId);

            if (idx > 0) {
                auto & next = tweets[u][idx - 1];
                pq.push({
                    next.first,
                    next.second,
                    idx - 1,
                    u
                });

            }
        }
        
        return res;

    }

    void follow(int followerId, int followeeId) {
        // write your code here 
        following[followerId].insert(followeeId);
    }

    void unfollow(int followerId, int followeeId) {
        // write your code here 
        following[followerId].erase(followeeId);

    }
};
